### PR TITLE
48897 NM duplicate files logic for weekly daily Files

### PR DESCRIPTION
**Pull request by Shirin14926**
Duplicate file logic code.
Unit test cases and Integration tests.

The reason behind adding this logic is there is no way to update existing batches for any modification/updation to weekly/daily files data, so there are chances that for same year & week, we may get multiple batches and in this case we should always take the latest batch details based on BatchPublishedDate.

### CODE REVIEW

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs [hunk 1]
```diff
@@ -34,15 +43,34 @@ public static List<ShowFilesResponseModel> ListFilesResponse(BatchSearchResponse
...
                 DataDate = item.Attributes.Where(x => x.Key.Equals("Data Date")).Select(x => x.Value).FirstOrDefault(),
                 WeekNumber = item.Attributes.Where(x => x.Key.Equals("Week Number")).Select(x => x.Value).FirstOrDefault(),
                 Year = item.Attributes.Where(x => x.Key.Equals("Year")).Select(x => x.Value).FirstOrDefault(),
-                YearWeek = item.Attributes.Where(x => x.Key.Equals("Year / Week")).Select(x => x.Value).FirstOrDefault(),
-                AllFilesZipSize = (long)item.AllFilesZipSize
+                YearWeek = item.Attributes.Where(x => x.Key.Replace(" ","").Equals("Year / Week".Replace(" ", ""))).Select(x => x.Value.Replace(" ", "")).FirstOrDefault(),
+                 AllFilesZipSize = (long)item.AllFilesZipSize
             }).ToList();
 
             IEnumerable<IGrouping<string, AttributesModel>> grouped = attributes.GroupBy(x => x.YearWeek);
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs:76]: Distinct attribute model for each date can be created using BatchPublishedDate as given below 'latestDailyfiles ':

Remember to add BatchPublishedDate in AttributesModel.

```suggestion
            IEnumerable<IGrouping<string, AttributesModel>> grouped = attributes.GroupBy(x => x.YearWeek);
             foreach (IGrouping<string, AttributesModel> group in grouped)
            {
                List<AttributesModel> latestDailyfiles = group.AsEnumerable()
                                                    .OrderByDescending(x => x.BatchPublishedDate)
                                                    .GroupBy(x => Convert.ToDateTime(x.DataDate))
                                                    .Select(x => x.First())
                                                    .ToList();

                List<DailyFilesDataModel> lstDataDate = (latestDailyfiles.Select(item => new DailyFilesDataModel
                {
```

**Review by usha13529 — CHANGES_REQUESTED** [review]
Few comments to fix please.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/FileShareService.cs [hunk 2]
```diff
@@ -22,7 +22,6 @@ public FileShareService(IOptions<FileShareServiceConfiguration> fileShareService
 
         public async Task<IResult<BatchSearchResponse>> FSSBatchSearchAsync(string searchText, string accessToken, string correlationId, IFileShareApiClient fileShareApiClient)
         {
-
             try
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/FileShareService.cs]: Removed, Thanks

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 3]
```diff
@@ -32,7 +32,7 @@ public async Task<List<ShowFilesResponseModel>> GetWeeklyBatchFiles(int year, in
             List<ShowFilesResponseModel> ListshowFilesResponseModels = new();
             try
             {
-                string accessToken = await _authFssTokenProvider.GenerateADAccessToken(correlationId);
+                string accessToken =await _authFssTokenProvider.GenerateADAccessToken(correlationId);
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: Removed, Thanks

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs [hunk 4]
```diff
@@ -34,15 +43,34 @@ public static List<ShowFilesResponseModel> ListFilesResponse(BatchSearchResponse
 
         public static List<ShowDailyFilesResponseModel> GetDailyShowFilesResponse(BatchSearchResponse searchResult)
         {
+            var distinctList = searchResult.Entries.Select(entry => new { YearWeek = entry.Attributes[5].Value, Date = entry.Attributes[0].Value, BatchPublishedDate = entry.BatchPublishedDate, BatchId = entry.BatchId }).OrderByDescending(x => x.BatchPublishedDate).ToList();
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs]: Refactor the code in single LINQ query. Thanks

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs [hunk 5]
```diff
@@ -34,15 +43,34 @@ public static List<ShowFilesResponseModel> ListFilesResponse(BatchSearchResponse
...
+                    {
+                        BatchDetailsList.Add(item);
+                        break;
+                    }
+                }
+            }
+            searchResult.Count = BatchDetailsList.Count;
+            searchResult.Total = BatchDetailsList.Count;
+            searchResult.Entries = BatchDetailsList;
+
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs]: Refactor the code in single LINQ query. Thanks

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs [hunk 6]
```diff
@@ -34,15 +43,34 @@ public static List<ShowFilesResponseModel> ListFilesResponse(BatchSearchResponse
...
             List<AttributesModel> attributes = searchResult.Entries.Where(x => x.AllFilesZipSize.HasValue).Select(item => new AttributesModel
             {
                 BatchId = item.BatchId,
                 DataDate = item.Attributes.Where(x => x.Key.Equals("Data Date")).Select(x => x.Value).FirstOrDefault(),
                 WeekNumber = item.Attributes.Where(x => x.Key.Equals("Week Number")).Select(x => x.Value).FirstOrDefault(),
                 Year = item.Attributes.Where(x => x.Key.Equals("Year")).Select(x => x.Value).FirstOrDefault(),
-                YearWeek = item.Attributes.Where(x => x.Key.Equals("Year / Week")).Select(x => x.Value).FirstOrDefault(),
-                AllFilesZipSize = (long)item.AllFilesZipSize
+                YearWeek = item.Attributes.Where(x => x.Key.Replace(" ","").Equals("Year / Week".Replace(" ", ""))).Select(x => x.Value.Replace(" ", "")).FirstOrDefault(),
+                 AllFilesZipSize = (long)item.AllFilesZipSize
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs]: Refactor the code in single LINQ query. Thanks

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs (hunk 1 above)
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs:61]: Done as expected. Thanks for suggesting this refactoring change.

**Review by usha13529 — APPROVED** [review]
Thanks for the changes @Shirin14926 . Looks good to me.

File: Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/NMDataServiceTest.cs [hunk 7]
```diff
@@ -514,5 +662,65 @@ private static Result<BatchAttributesSearchResponse> SetAttributeSearchNoYearWee
             };
             return attributeSearchResult;
         }
+
+        private static Result<BatchSearchResponse> SetSearchResultForWeeklyDuplicateFile()
```
Comment (barrie-cooper) [inline Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/NMDataServiceTest.cs:666]: Just to be consistent with `SetSearchResultForDuplicateDailyFiles`
```suggestion
        private static Result<BatchSearchResponse> SetSearchResultForDuplicateWeeklyFiles()
```

**Review by barrie-cooper — COMMENTED** [review]
A few very minor suggestions

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs [hunk 8]
```diff
@@ -8,9 +8,18 @@ namespace UKHO.MaritimeSafetyInformation.Common.Helpers
...
+            {
+                List<BatchDetails> BatchDetailsList = new();
+
+                BatchDetails batchDetail = SearchResult.Entries.OrderByDescending(t => t.BatchPublishedDate).FirstOrDefault();
+                BatchDetailsList.Add(batchDetail);
+                SearchResult.Entries = BatchDetailsList;
+                SearchResult.Count = BatchDetailsList.Count;
+                SearchResult.Total = BatchDetailsList.Count;
+            }
             List<ShowFilesResponseModel> ListshowFilesResponseModels = new();
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs]: done, Thanks

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs [hunk 9]
```diff
@@ -34,16 +43,20 @@ public static List<ShowFilesResponseModel> ListFilesResponse(BatchSearchResponse
...
             List<AttributesModel> attributes = searchResult.Entries.Where(x => x.AllFilesZipSize.HasValue).Select(item => new AttributesModel
             {
                 BatchId = item.BatchId,
                 DataDate = item.Attributes.Where(x => x.Key.Equals("Data Date")).Select(x => x.Value).FirstOrDefault(),
                 WeekNumber = item.Attributes.Where(x => x.Key.Equals("Week Number")).Select(x => x.Value).FirstOrDefault(),
                 Year = item.Attributes.Where(x => x.Key.Equals("Year")).Select(x => x.Value).FirstOrDefault(),
-                YearWeek = item.Attributes.Where(x => x.Key.Equals("Year / Week")).Select(x => x.Value).FirstOrDefault(),
-                AllFilesZipSize = (long)item.AllFilesZipSize
-            }).ToList();
+                YearWeek = item.Attributes.Where(x => x.Key.Replace(" ", "").Equals("Year / Week".Replace(" ", ""))).Select(x => x.Value.Replace(" ", "")).FirstOrDefault(),
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs]: Done, Thanks

File: Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/NMDataServiceTest.cs (hunk 7 above)
Comment (Shirin14926) [reply Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/NMDataServiceTest.cs]: Done, Thanks

File: Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/NMDataServiceTest.cs [hunk 10]
```diff
@@ -114,7 +114,43 @@ public async Task WhenGetDailyBatchDetailsFilesIsCalled_ThenShouldReturnsMoreTha
...
+        public async Task WhenGetDailyBatchDetailsFilesIsCalledWithDuplicateData_ThenShouldReturnLatestFiles()
+        {
+            A.CallTo(() => _fakeAuthFssTokenProvider.GenerateADAccessToken(A<string>.Ignored));
+
+            Result<BatchSearchResponse> searchResult = SetSearchResultForDuplicateDailyFiles();
+
+            A.CallTo(() => _fakefileShareService.FSSBatchSearchAsync(A<string>.Ignored, A<string>.Ignored, A<string>.Ignored, A<IFileShareApiClient>.Ignored)).Returns(searchResult);
+
+            const int expectedRecordCount = 1;
+            const int dailyFilesDataCount = 2;
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/NMDataServiceTest.cs]: Done, Thanks

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs [hunk 11]
```diff
@@ -8,9 +8,18 @@ namespace UKHO.MaritimeSafetyInformation.Common.Helpers
 {
     public static class NMHelper
     {
-
         public static List<ShowFilesResponseModel> ListFilesResponse(BatchSearchResponse SearchResult)
         {
+            if (SearchResult.Entries.Count > 1)
+            {
+                List<BatchDetails> BatchDetailsList = new();
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs]: renamed, thanks

**Review by barrie-cooper — APPROVED** [review]
LGTM, thanks @Shirin14926

**Pull request merged**